File: uikit/src/main/java/com/sendbird/uikit/widgets/EmojiReactionUserListView.java
package com.sendbird.uikit.widgets;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import com.sendbird.android.Reaction;
import com.sendbird.android.User;
import com.sendbird.uikit.R;
import com.sendbird.uikit.activities.adapter.EmojiReactionUserListAdapter;
import com.sendbird.uikit.databinding.SbFragmentUserListBinding;
import com.sendbird.uikit.databinding.SbViewEmojiReactionUserListBinding;
import com.sendbird.uikit.model.EmojiManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class EmojiReactionUserListView extends FrameLayout {
    private SbViewEmojiReactionUserListBinding binding;

    public EmojiReactionUserListView(@NonNull Context context) {
        this(context, null);
    }

    public EmojiReactionUserListView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.sb_widget_emoji_message);
    }

    public EmojiReactionUserListView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    public void init(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.EmojiReactionCountList, defStyleAttr, R.style.Widget_Sendbird_Emoji);

        try {
            binding = SbViewEmojiReactionUserListBinding.inflate(LayoutInflater.from(context), this, true);
            int tabLayoutBackgroundResId = a.getResourceId(R.styleable.EmojiReactionCountList_sb_emoji_reaction_count_tab_layout_background, R.drawable.sb_tab_layout_border_background_light);
            int indicatorColor = a.getColor(R.styleable.EmojiReactionCountList_sb_emoji_reaction_count_tab_indicator_color, context.getResources().getColor(R.color.primary_300));
            binding.tabLayoutPanel.setBackgroundResource(tabLayoutBackgroundResId);
            binding.tabLayout.setSelectedTabIndicatorColor(indicatorColor);
        } finally {
            a.recycle();
        }
    }

    public void setEmojiReactionUserData(@NonNull Fragment fragment, int currentPosition, @NonNull List<Reaction> reactionList, @NonNull Map<Reaction, List<User>> reactionUserInfo) {
        UserListPagerAdapter pagerAdapter = new UserListPagerAdapter(getContext(), fragment, reactionList, reactionUserInfo);
        binding.vpEmojiReactionUserList.setAdapter(pagerAdapter);
        new TabLayoutMediator(binding.tabLayout, binding.vpEmojiReactionUserList,
                (tab, position) -> {
                    EmojiReactionCountView view = new EmojiReactionCountView(getContext());
                    Reaction reaction = reactionList.get(position);
                    if (reaction != null && reaction.getUserIds() != null) {
                        view.setCount(reaction.getUserIds().size());
                        view.setEmojiUrl(EmojiManager.getInstance().getEmojiUrl(reaction.getKey()));
                    }
                    tab.setCustomView(view);
                }
        ).attach();

        TabLayout.Tab defaultTab = binding.tabLayout.getTabAt(currentPosition);
        if (defaultTab != null) {
            defaultTab.select();
        }
    }

    private static class UserListPagerAdapter extends FragmentStateAdapter {
        private final int itemCount;
        private final List<Fragment> fragmentList = new ArrayList<>();

        UserListPagerAdapter(@NonNull Context context,
                             @NonNull Fragment fragment,
                             @NonNull List<Reaction> reactionList,
                             @NonNull Map<Reaction, List<User>> reactionUserInfo) {
            super(fragment);
            this.itemCount = reactionUserInfo.size();
            for (Reaction reaction: reactionList) {
                final List<User> userList = reactionUserInfo.get(reaction);
                fragmentList.add(new UserListFragment(context, userList == null ? Collections.emptyList(): userList));
            }
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            return fragmentList.get(position);
        }

        @Override
        public int getItemCount() {
            return itemCount;
        }
    }

    public static class UserListFragment extends Fragment {
        private final Context context;
        private SbFragmentUserListBinding binding;
        private final List<User> userList;

        UserListFragment(@NonNull Context context, @NonNull List<User> userList) {
            this.context = context;
            this.userList = new ArrayList<>(userList);
        }

        @Override
        @NonNull
        public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                                 @Nullable Bundle savedInstanceState) {
            LayoutInflater themeInflater = inflater.cloneInContext(context);
            binding = SbFragmentUserListBinding.inflate(themeInflater);
            binding.rvUserList.setUseDivider(false);
            return binding.getRoot();
        }

        @Override
        public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            EmojiReactionUserListAdapter userListAdapter = new EmojiReactionUserListAdapter(userList);
            binding.rvUserList.setAdapter(userListAdapter);
            binding.rvUserList.setHasFixedSize(true);
        }
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/ui/reactions/EmojiReactionUserListView.kt
package com.sendbird.uikit.internal.ui.reactions

import android.content.Context
import android.os.Bundle
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.sendbird.android.message.Reaction
import com.sendbird.android.user.User
import com.sendbird.uikit.R
import com.sendbird.uikit.activities.adapter.EmojiReactionUserListAdapter
import com.sendbird.uikit.consts.StringSet
import com.sendbird.uikit.databinding.SbFragmentUserListBinding
import com.sendbird.uikit.databinding.SbViewEmojiReactionUserListBinding
import com.sendbird.uikit.interfaces.OnItemClickListener
import com.sendbird.uikit.model.EmojiManager

internal class EmojiReactionUserListView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = R.attr.sb_widget_emoji_message
) : FrameLayout(context, attrs, defStyleAttr) {
    private val binding: SbViewEmojiReactionUserListBinding

    var onProfileClickListener: OnItemClickListener<User>? = null

    init {
        val a = context.theme.obtainStyledAttributes(
            attrs,
            R.styleable.EmojiReactionCountList,
            defStyleAttr,
            R.style.Widget_Sendbird_Emoji
        )
        try {
            binding = SbViewEmojiReactionUserListBinding.inflate(LayoutInflater.from(context), this, true)
            val tabLayoutBackgroundResId = a.getResourceId(
                R.styleable.EmojiReactionCountList_sb_emoji_reaction_count_tab_layout_background,
                R.drawable.sb_tab_layout_border_background_light
            )
            val indicatorColor = a.getColor(
                R.styleable.EmojiReactionCountList_sb_emoji_reaction_count_tab_indicator_color,
                ContextCompat.getColor(context, R.color.primary_300)
            )
            binding.tabLayoutPanel.setBackgroundResource(tabLayoutBackgroundResId)
            binding.tabLayout.setSelectedTabIndicatorColor(indicatorColor)
        } finally {
            a.recycle()
        }
    }

    fun setEmojiReactionUserData(
        fragment: Fragment,
        currentPosition: Int,
        reactionList: List<Reaction>,
        reactionUserInfo: Map<Reaction, List<User?>>
    ) {
        val pagerAdapter = UserListPagerAdapter(fragment, reactionList, reactionUserInfo).apply {
            this.onProfileClickListener = OnItemClickListener<User> { view, position, data ->
                this@EmojiReactionUserListView.onProfileClickListener?.onItemClick(view, position, data)
            }
        }
        binding.vpEmojiReactionUserList.adapter = pagerAdapter
        TabLayoutMediator(binding.tabLayout, binding.vpEmojiReactionUserList) { tab: TabLayout.Tab, position: Int ->
            val view = EmojiReactionCountView(context)
            val reaction = reactionList[position]
            view.setCount(reaction.userIds.size)
            view.setEmojiUrl(EmojiManager.getEmojiUrl(reaction.key))
            tab.customView = view
        }.attach()
        val defaultTab = binding.tabLayout.getTabAt(currentPosition)
        defaultTab?.select()
    }

    private class UserListPagerAdapter(
        fragment: Fragment,
        reactionList: List<Reaction>,
        reactionUserInfo: Map<Reaction, List<User?>>
    ) : FragmentStateAdapter(fragment) {
        private val itemCount: Int
        private val fragmentList: MutableList<Fragment> = ArrayList()
        var onProfileClickListener: OnItemClickListener<User> = OnItemClickListener<User> { _, _, _ -> }

        override fun createFragment(position: Int): Fragment = fragmentList[position]
        override fun getItemCount(): Int = itemCount

        init {
            itemCount = reactionUserInfo.size
            reactionList.forEach {
                val userListFragment = UserListFragment().apply {
                    this.onProfileClickListener = OnItemClickListener<User> { view, position, data ->
                        this@UserListPagerAdapter.onProfileClickListener.onItemClick(view, position, data)
                    }
                }
                reactionUserInfo[it]?.let { userList ->
                    val bundle = Bundle()
                    bundle.putInt(StringSet.KEY_EMOJI_REACTION_USER_LIST_SIZE, userList.size)
                    userList.forEachIndexed { index, user ->
                        bundle.putByteArray(StringSet.KEY_EMOJI_REACTION_USER_ + index.toString(), user?.serialize())
                    }
                    userListFragment.arguments = bundle
                }
                fragmentList.add(userListFragment)
            }
        }
    }

    // A class that inherits the Fragment must be public static.
    // If user who was already reacted has been banned or was deactivated, the reacted user may be empty.
    internal class UserListFragment : Fragment() {
        private lateinit var binding: SbFragmentUserListBinding
        var onProfileClickListener: OnItemClickListener<User> = OnItemClickListener<User> { _, _, _ -> }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
            val themeInflater = inflater.cloneInContext(context)
            binding = SbFragmentUserListBinding.inflate(themeInflater)
            binding.rvUserList.setUseDivider(false)
            return binding.root
        }

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            val userList = mutableListOf<User?>()
            arguments?.let {
                val userListSize = it.getInt(StringSet.KEY_EMOJI_REACTION_USER_LIST_SIZE)
                for (i in 0 until userListSize) {
                    val user =
                        User.buildFromSerializedData(it.getByteArray(StringSet.KEY_EMOJI_REACTION_USER_ + i.toString()))
                    userList.add(user)
                }
            }
            val userListAdapter = EmojiReactionUserListAdapter(userList)
            userListAdapter.setOnEmojiReactionUserListProfileClickListener { v, position, item ->
                onProfileClickListener.onItemClick(
                    v,
                    position,
                    item
                )
            }
            binding.rvUserList.adapter = userListAdapter
            binding.rvUserList.setHasFixedSize(true)
        }
    }
}
